### PR TITLE
Add custom delimiter attribute for templating

### PR DESCRIPTION
**Pull request by alabiaga**
- add validator support for new `custom-delimiters` attribute to `<template>` and `<script>` templating tags. Intentionally limited to AMP, AMP4ADS and ACTIONS html format.
- add `#-,-#` as an option to custom delimiters, rather than default mustache `{,}`. NOTE that `#-,-#`  is an arbritrary beginning and ending delimiter. erb style delimiter `<% example %>` is the most popular that i've seen but it is not something we want to support, for context: https://github.com/ampproject/amphtml/issues/12994#issuecomment-480135247

Implementation: #24012

### CODE REVIEW

**Comment by alabiaga** [comment]
cc\ @zhangsu @GoTcWang

**Comment by alabiaga** [comment]
cc\ @zhangsu @GoTcWang

**Comment by choumx** [comment]
@jridgewell @Gregable I don't quite recall the parsing issue with `<%` and `%>`. It can't be that the validator's parser doesn't parse those correctly right? 

https://html.spec.whatwg.org/multipage/parsing.html#tag-open-state

**Comment by jridgewell** [comment]
Using `<%` (anything with `<`) was rejected because other parsers may not interpret them correctly.

**Comment by choumx** [comment]
I think we can safely ignore parsers that don't conform to such a fundamental HTML parsing rule.

**Comment by alabiaga** [comment]
> I think we can safely ignore parsers that don't conform to such a fundamental HTML parsing rule.

Some of the publishers who brought up the usefulness of custom delimiters also gave code samples which showed usage of the ERB style so I too am in favor of allowing it. How can we move forward with this change? I would think that is up to publisher to use their discretion in their choice of the custom delimiter and the parser they use for their other templating needs.

**Comment by alabiaga** [comment]
Friendly ping. Any way I can help move this forward. This PR will be harder to merge as it gets stale.

**Comment by alabiaga** [comment]
cc\ @zhangsu @GoTcWang

File: extensions/amp-mustache/0.2/test/validator-amp-mustache.html [hunk 1]
```diff
@@ -132,6 +132,16 @@
...
 </template>
 
+<!-- Template with allowed custom delimiters -->
+<template type="amp-mustache" custom-delimiters="#- -#">
+  Hello #-name-#
+</template>
+
+<!-- Template with unsupported custom delimiters -->
+<template type="amp-mustache" custom-delimiters="<% %>">
+    Hello #-name-#
```
Comment (Gregable) [inline extensions/amp-mustache/0.2/test/validator-amp-mustache.html]: should this match the attribute in your example?

File: extensions/amp-mustache/validator-amp-mustache.protoascii [hunk 2]
```diff
@@ -75,6 +75,10 @@ tags: {
     value: "amp-mustache"
     dispatch_key: NAME_VALUE_DISPATCH
   }
+  attrs: {
+    name: "custom-delimiters"
+    value_regex: "#- -#"
```
Comment (Gregable) [inline extensions/amp-mustache/validator-amp-mustache.protoascii]: can this just be `value:`? why regex?

File: extensions/amp-mustache/validator-amp-mustache.protoascii [hunk 3]
```diff
@@ -270,6 +274,10 @@ tags: {
     mandatory: true
     value: "amp-mustache"
   }
+  attrs: {
+    name: "custom-delimiters"
+    value_regex: "#- -#"
```
Comment (Gregable) [inline extensions/amp-mustache/validator-amp-mustache.protoascii]: ditto

File: extensions/amp-mustache/validator-amp-mustache.protoascii (hunk 2 above)
Comment (choumx) [inline extensions/amp-mustache/validator-amp-mustache.protoascii]: Really not a fan of `#- -#`. How about `[[ ]]` instead?

**Comment by Gregable** [comment]
The validator actually does a lot of validation within `<template>` tags, since those are actual HTML parsed by the DOM before mustache runs. `{{` is hard coded for this validation. This isn't exercised by the simple tests you added.

This validation messages will be poor for your use case. For example:

```
<p {{notallowed}}></p>
   ^ Mustache template syntax in attribute name '{{notallowed}}' in tag 'p'.
```

vs. 

```
<p #-notallowed-#></p>
   ^ The attribute '#-notallowed-#' may not appear in tag 'p'.
```

Also, in cases where the rules have specific requirements, this syntax likely won't even pass validation. For example:

```
<a href="" target="{{foo}}"></a> 
```

No error.


```
<a href="" target="#-foo-#"></p>
                   ^The attribute 'target' in tag 'a' is set to the invalid value '#-foo-#'.
```

Error.

**Comment by alabiaga** [comment]
@choumx @Gregable thanks for the feedback. I'll make the changes and will reach out to David for the C++ changes. 

cc/ @honeybadgerdontcare

**Comment by zhangsu** [comment]
+1 to Will. Please make sure to disable this feature for AMP for Email until various email clients support custom delimiters. In addition to validator rules, there are work on the server-side too to support custom delimiter (e.g., server-side rendering of Mustache templates).

**Comment by alabiaga** [comment]
> +1 to Will. Please make sure to disable this feature for AMP for Email until various email clients support custom delimiters. In addition to validator rules, there are work on the server-side too to support custom delimiter (e.g., server-side rendering of Mustache templates).

Right, yes I captured that in the internal design doc. I will make this change. Thanks for the reminder.

**Comment by mixxor** [comment]
Hey There, thanks for the nice work on AMP,
any Chance we can have this?

Working on vue.js + AMP Mustache is a pain without it.

**Comment by mrjoro** [comment]
Any updates on this feature @choumx?

**Pull request closed without merging**